app/src/main/java/com/google/android/sambadocumentsprovider/browsing/broadcast/BroadcastUtils.java

```diff
@@ -0,0 +1,162 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.google.android.sambadocumentsprovider.browsing.broadcast;
+
+import android.util.Log;
+
+import com.google.android.sambadocumentsprovider.browsing.BrowsingException;
+
+import java.net.InetAddress;
+import java.net.InterfaceAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+
+class BroadcastUtils {
+  private static final int FILE_SERVER_NODE_TYPE = 0x20;
+
+  /**
+   * Generates a NetBIOS name query request.
+   * https://tools.ietf.org/html/rfc1002
+   * Section 4.2.12
+   */
+  static byte[] createPacket() {
+    ByteBuffer os = ByteBuffer.allocate(50);
+
+    char transId = 5566;
+    char broadcastFlag = 0x0010;
+    char questionCount = 1;
+    char answerResourceCount = 0;
+    char authorityResourceCount = 0;
+    char additionalResourceCount = 0;
+
+    os.putChar(transId);
+    os.putChar(broadcastFlag);
+    os.putChar(questionCount);
+    os.putChar(answerResourceCount);
+    os.putChar(authorityResourceCount);
+    os.putChar(additionalResourceCount);
+
+    // Length of name. 16 bytes of name encoded to 32 bytes.
+    os.put((byte) 0x20);
+
+    // '*' character encodes to 2 bytes.
+    os.put((byte) 0x43);
+    os.put((byte) 0x4b);
+
+    // Write the remaining 15 nulls which encode to 30* 0x41
+    for (int i = 0; i < 30; i++) {
+      os.put((byte) 0x41);
+    }
+
+    // Length of next segment.
+    os.put((byte) 0);
+
+    // Question type: Node status
+    os.putChar((char) 0x21);
+
+    // Question class: Internet
+    os.putChar((char) 0x01);
+
+    return os.array();
+  }
+
+  /**
+   * Parses a positive response to NetBIOS name request query.
+   * https://tools.ietf.org/html/rfc1002
+   * Section 4.2.13
+   */
+  static List<String> extractServers(byte[] data) throws BrowsingException {
+    ByteBuffer buffer = ByteBuffer.wrap(data);
+
+    skipBytes(buffer, 2); // Skip transId.
+    skipBytes(buffer, 2); // Skip flags.
+
+    skipBytes(buffer, 2); // No questions.
+    skipBytes(buffer, 2); // Skip answers count.
+    skipBytes(buffer, 2); // No authority resources.
+    skipBytes(buffer, 2); // No additional resources.
+
+    int nameLength = buffer.get();
+    skipBytes(buffer, nameLength);
+
+    skipBytes(buffer, 1);
+
+    int nodeStatus = buffer.getChar();
+    if (nodeStatus != 0x20 && nodeStatus != 0x21) {
+      throw new BrowsingException("Received negative response for the broadcast");
+    }
+
+    skipBytes(buffer, 2);
+    skipBytes(buffer, 4);
+    skipBytes(buffer, 2);
+
+    int addressListEntryCount = buffer.get();
+
+    List<String> servers = new ArrayList<>();
+
+    for (int i = 0; i < addressListEntryCount; i++) {
+      StringBuilder nameBuilder = new StringBuilder();
+      for (int j = 0; j < 15; j++) {
```

Comment:
I think instead of this loop you can have a byte array and use the ByteBuffer.get() that writes into an array. Then use the string constructor that take a byte array and specify  the charset to be ASCII. Then you won't need a stringbulider.